### ISSUE TITLE
bazel: build fails with "could not find dependency" npm error

### ISSUE THREAD

**Issue by aliher1911** [cockroachdb/cockroach#70867]
**Describe the problem**

When running ./dev babel generate I'm getting back following error:

```
Loading: 
Loading: 0 packages loaded
Analyzing: target //:gazelle (0 packages loaded, 0 targets configured)
INFO: Analyzed target //:gazelle (1 packages loaded, 390 targets configured).
INFO: Found 1 target...
WARNING: failed to create one or more convenience symlinks for prefix '_bazel/':
  cannot create symbolic link bazel-out -> /private/var/tmp/_bazel_user/3de4b0d5f2cd1e24107aaf9b3a0a01a6/execroot/cockroach/bazel-out:  /Users/user/go/src/github.com/cockroachdb/cockroach/bazel-out (File exists)
[0 / 2] [Prepa] BazelWorkspaceStatusAction stable-status.txt
Target //:gazelle up-to-date:
  _bazel/bin/gazelle-runner.bash
  _bazel/bin/gazelle
INFO: Elapsed time: 1.037s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: _bazel/bin/gazelle update-repos '-from_file=go.mod' '-build_file_proto_mode=disable_global' '-to_macro=DEPS.bzl%go_deps' '-prune=true'
INFO: Build completed successfully, 1 total action
Loading: 
Loading: 0 packages loaded
INFO: Build option --run_under has changed, discarding analysis cache.
Analyzing: target //pkg/cmd/generate-test-suites:generate-test-suites (0 packages loaded, 0 targets configured)
INFO: Analyzed target //pkg/cmd/generate-test-suites:generate-test-suites (0 packages loaded, 7402 targets configured).
INFO: Found 1 target...
WARNING: failed to create one or more convenience symlinks for prefix '_bazel/':
  cannot create symbolic link bazel-out -> /private/var/tmp/_bazel_user/3de4b0d5f2cd1e24107aaf9b3a0a01a6/execroot/cockroach/bazel-out:  /Users/user/go/src/github.com/cockroachdb/cockroach/bazel-out (File exists)
[0 / 4] [Prepa] BazelWorkspaceStatusAction stable-status.txt
Target //pkg/cmd/generate-test-suites:generate-test-suites up-to-date:
  _bazel/bin/pkg/cmd/generate-test-suites/generate-test-suites_/generate-test-suites
INFO: Elapsed time: 1.223s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: /bin/bash -c 'cd /Users/user/go/src/github.com/cockroachdb/cockroach &&  _bazel/bin/pkg/cmd/generate-test-suites/generate-test-suites_/generate-test-suites '
INFO: Build completed successfully, 1 total action
2021/09/29 16:12:31 Could not query Bazel tests: got error exit status 7
2021/09/29 16:12:31 Got error output: Loading: 0 packages loaded
Loading: 1 packages loaded
    currently loading: pkg/ui/workspaces/db-console ... (4 packages)
Loading: 1 packages loaded
    currently loading: pkg/ui/workspaces/db-console ... (4 packages)
Loading: 1 packages loaded
    currently loading: pkg/ui/workspaces/db-console ... (4 packages)
INFO: Repository npm instantiated at:
  /Users/user/go/src/github.com/cockroachdb/cockroach/WORKSPACE:67:13: in <toplevel>
  /private/var/tmp/_bazel_user/3de4b0d5f2cd1e24107aaf9b3a0a01a6/external/build_bazel_rules_nodejs/index.bzl:83:18: in yarn_install
Repository rule yarn_install defined at:
  /private/var/tmp/_bazel_user/3de4b0d5f2cd1e24107aaf9b3a0a01a6/external/build_bazel_rules_nodejs/internal/npm_install/npm_install.bzl:890:31: in <toplevel>
Loading: 1 packages loaded
    currently loading: pkg/ui/workspaces/db-console ... (4 packages)
ERROR: An error occurred during the fetch of repository 'npm':
   Traceback (most recent call last):
	File "/private/var/tmp/_bazel_user/3de4b0d5f2cd1e24107aaf9b3a0a01a6/external/build_bazel_rules_nodejs/internal/npm_install/npm_install.bzl", line 888, column 24, in _yarn_install_impl
		_create_build_files(repository_ctx, "yarn_install", node, repository_ctx.attr.yarn_lock, repository_ctx.attr.generate_local_modules_build_files)
	File "/private/var/tmp/_bazel_user/3de4b0d5f2cd1e24107aaf9b3a0a01a6/external/build_bazel_rules_nodejs/internal/npm_install/npm_install.bzl", line 465, column 13, in _create_build_files
		fail("generate_build_file.ts failed: \nSTDOUT:\n%s\nSTDERR:\n%s" % (result.stdout, result.stderr))
Error in fail: generate_build_file.ts failed: 
STDOUT:

STDERR:
could not find dependency '@babel/helper-define-map' of 'babel-preset-react-app/node_modules/@babel/preset-env/node_modules/@babel/plugin-transform-classes'
ERROR: Error fetching repository: Traceback (most recent call last):
	File "/private/var/tmp/_bazel_user/3de4b0d5f2cd1e24107aaf9b3a0a01a6/external/build_bazel_rules_nodejs/internal/npm_install/npm_install.bzl", line 888, column 24, in _yarn_install_impl
		_create_build_files(repository_ctx, "yarn_install", node, repository_ctx.attr.yarn_lock, repository_ctx.attr.generate_local_modules_build_files)
	File "/private/var/tmp/_bazel_user/3de4b0d5f2cd1e24107aaf9b3a0a01a6/external/build_bazel_rules_nodejs/internal/npm_install/npm_install.bzl", line 465, column 13, in _create_build_files
		fail("generate_build_file.ts failed: \nSTDOUT:\n%s\nSTDERR:\n%s" % (result.stdout, result.stderr))
Error in fail: generate_build_file.ts failed: 
STDOUT:

STDERR:
could not find dependency '@babel/helper-define-map' of 'babel-preset-react-app/node_modules/@babel/preset-env/node_modules/@babel/plugin-transform-classes'
ERROR: generate_build_file.ts failed: 
STDOUT:

STDERR:
could not find dependency '@babel/helper-define-map' of 'babel-preset-react-app/node_modules/@babel/preset-env/node_modules/@babel/plugin-transform-classes'
Loading: 1 packages loaded
    currently loading: pkg/ui/workspaces/db-console ... (4 packages)
Loading: 1 packages loaded
    currently loading: pkg/ui/workspaces/db-console ... (4 packages)
ERROR: exit status 1

```

using make to build cockroach works fine with/without ui.

**To Reproduce**

Using commit: 55e2bcd4affa844b8288da23407649d8f52a1e23
run
`./dev generate bazel
`

Building under macos, using bazelisk.

Epic CRDB-8036

**Comment by rafiss** [comment]
I'm encountering this too now

**Comment by rickystewart** [comment]
Pull from `master` and you shouldn't see the `dev generate bazel` issue any longer. You'll still see the same when you `dev build short` though.

**Comment by rickystewart** [comment]
The `dev build short` issue will remain until #70828 is merged, but the RC is still unknown.

@koorosh Does this build error mean anything to you?

**Comment by koorosh** [comment]
hm, couldn't reproduce from the same commits (e8634f9, or 55e2bcd).

> ERROR: Error fetching repository: Traceback (most recent call last):
and this error shows up when NPM package couldn't be loaded from network 🤔

**Comment by irfansharif** [comment]
@andreimatei claims a `make ui-maintainer-clean` got him past the error.

**Comment by rickystewart** [comment]
I suspect that `rules_nodejs` can get the `node_modules` folders into a bad state that would require cleaning, but I don't have any way to validate that inkling.

We can add something like `dev clean --ui` as an alternative to `make ui-maintainer-clean`, which would help `dev` usability if nothing else.

**Comment by irfansharif** [comment]
See this [internal discussion](https://cockroachlabs.slack.com/archives/C019A8NANH0/p1643647779106099). Found this easy-ish to repro: run `make build` and `dev build` in quick succession, sometimes just aborting the `make` process part way through.

TL;DR the `rules_nodejs` library we're using prepares `node_modules` in the source tree itself, which if run concurrently with `make` (also writes to the same place) might cause residual badness. In addition to `make ui-maintainer-clean`, `bazel clean` will help. See [this](
https://github.com/bazelbuild/rules_nodejs/pull/2946/files#diff-8905d32b7906216ecf1fc9d732c6895cfe17f7274216aefd32482516d1764ed8R1133-R1137) from upstream.

How should we improve this? Bazel writing to the source tree itself feels a bit odd. More recent versions of `rules_nodejs` have moved away from the pattern: https://bazelbuild.github.io/rules_nodejs/dependencies.html#symlink_node_modules-and-managed_directories. I tried opting out of the symlink mode and bumping rules_nodejs, but didn't get far. I settled with the simpler attempt to just have bazel and make write to different paths when crafting `node_modules`. 

https://gist.github.com/irfansharif/58f8c347db49765839feff9e563d07a4

Probably unfortunate to have to maintain two package.json + yarn.lock; I'm not sure how these files are maintained. I'll leave this issue to @koorosh, who may have other ideas to make this less painful.
